### PR TITLE
[spinel] add RCP API Version and compatibility check

### PR DESCRIPTION
**Pull request by abtink**
This PR contains two related commits:

**[spinel] add RCP API Version and compatibility check**

This commit adds a new compatibility check model for RCP and host. It
updates the compatibility guideline documentation stating how the
compatibility policy for NCP versus RCP are handled differently. For
NCP, we continue to require new NCP firmware to remain backward
compatible and work with an older host driver.

For RCP and host, this commit introduces the RCP API Version model.
The API version numbers are used to check the compatibility between
host implementation and RCP firmware. Generally, a newer host side
implementation would work with a range of previous or older RCP
firmware versions.

This commit adds `SPINEL_RCP_API_VERSION` definition which  specifies
the current spinel RCP API version. This number MUST be incremented
anytime there is a change in any of RCP specific spinel definitions.
`SPINEL_MIN_HOST_SUPPORTED_RCP_API_VERSION` specifies the minimum
spinel RCP API version which is supported by the host-side
implementation.

This commit also reserves a block of spinel properties value ranges
from `0xB0 - 0xFF` for RCP related property definitions. The smaller
property values help decrease the encoded byte stream length,
resulting in shorter spinel frames exchanged between RCP and host.

---------

**[radio-spinel] check RCP API Version to be in the supported range**

This commit adds a new method `CheckRcpApiVersion()` which is used
during `Init()` to get the API version from RCP and verify that it
is within supported range.

### CODE REVIEW

**Review by bukepo — APPROVED** [review]
LGTM:+1:

**Review by brianljt — APPROVED** [review]
Looks great. Thank you! 👍

File: src/lib/spinel/spinel.h [hunk 1]
```diff
@@ -305,6 +308,24 @@
...
+ *   compatibility between host implementation and RCP firmware. Generally,
+ *   a newer host side implementation would work with a range of previous
+ *   or older RCP firmware versions.
+ *
+ *   - SPINEL_RCP_API_VERSION specifies the current spinel RCP API version.
+ *     This number MUST be incremented anytime there is a change in any of RCP
+ *     specific spinel definitions.
+ *
+ *   - SPINEL_MIN_HOST_SUPPORTED_RCP_API_VERSION specifies the minimum spinel
+ *      RCP API Version which is supported by the host-side implementation.
```
Comment (jwhui) [inline src/lib/spinel/spinel.h:323]:
```suggestion
 *   - SPINEL_MIN_HOST_SUPPORTED_RCP_API_VERSION specifies the minimum spinel
 *     RCP API Version which is supported by the host-side implementation.
```

File: src/lib/spinel/spinel.h [hunk 2]
```diff
@@ -305,6 +308,24 @@
...
+ *
+ *   - On start host implementation would query the RCP API version and accept
```
Comment (jwhui) [inline src/lib/spinel/spinel.h:325]:
```suggestion
 *   - On start, host implementation queries the RCP API version and accepts
```

File: src/lib/spinel/spinel.h [hunk 3]
```diff
@@ -305,6 +308,24 @@
...
+ *     any version number from SPINEL_MIN_HOST_SUPPORTED_RCP_API_VERSION up to
+ *     SPINEL_RCP_API_VERSION.
```
Comment (jwhui) [inline src/lib/spinel/spinel.h:327]:
```suggestion
 *     and including SPINEL_RCP_API_VERSION.
```

**Review by jwhui — APPROVED** [review]

**Pull request merged**